lib/dialects/postgres/query-generator.js

```diff
@@ -357,7 +357,7 @@ const QueryGenerator = {
 
   upsertQuery(tableName, insertValues, updateValues, where, model, options) {
     const insert = this.insertQuery(tableName, insertValues, model.rawAttributes, options);
-    const update = this.updateQuery(tableName, updateValues, where, options, model.rawAttributes);
+    const update = this.updateQuery(tableName, updateValues, where, Object.assign({ model }, options), model.rawAttributes);
```

Comment:
I think `Model.upsert` should set it, like other methods do example [destroy](https://github.com/sequelize/sequelize/blob/6adfabf1d6f3bb9986bfebba4d966228f24e553f/lib/model.js#L2463)

Please add `options.model = this` around [here](6adfabf1d6f3bb9986bfebba4d966228f24e553f) and remove this change.